### PR TITLE
Fix "stdscr" configure failure & ncursest support

### PR DESCRIPTION
**Pull request by Explorer09**
This fixes the issue of #1548, while adding minor improvements to the ncurses detecttion code since #1512.

### CODE REVIEW

**Review by BenBE — CHANGES_REQUESTED** [review]
LGTM.

Document the `#if defined(HAVE_CONFIG_H)` to be used for reference inside `configure.ac`.

Also, if not done, have the ncursest specific changes re macros and the `ProvideCurses` changes live inside separate commits (though that already seems to be the case AFAICS).

File: configure.ac [hunk 1]
```diff
@@ -68,6 +68,15 @@ esac
 # Enable extensions, required by hwloc scripts
 AC_USE_SYSTEM_EXTENSIONS
 
+dnl May be needed for htop Unicode support
+AH_VERBATIM([_XOPEN_SOURCE_EXTENDED],
+[/* Enables XPG4v2 (SUSv1) interfaces if they are not enabled already
+   with _XOPEN_SOURCE */
+#ifndef _XOPEN_SOURCE_EXTENDED
+#undef _XOPEN_SOURCE_EXTENDED
+#endif
```
Comment (BenBE) [inline configure.ac:77]: I think you got the check inverted by accident …

```suggestion
#ifdef _XOPEN_SOURCE_EXTENDED
#undef _XOPEN_SOURCE_EXTENDED
#endif
```

File: configure.ac [hunk 2]
```diff
@@ -665,6 +684,28 @@ else
...
+#if defined(HAVE_LIBNCURSESW)
+{
+   static cchar_t dummy;
+   mvadd_wchnstr(0, 0, &dummy, 0);
+}
+#endif
+   ]])],
+   [AC_MSG_RESULT(yes)],
+   [AC_MSG_RESULT(no)
+   AC_MSG_FAILURE([there are problems with the curses header])])
```
Comment (BenBE) [inline configure.ac:708]:
```suggestion
   AC_MSG_FAILURE([there are problems with the curses header])
])
```

File: configure.ac (hunk 1 above)
Comment (BenBE) [reply configure.ac:77]: TIL. Please add a small note about this to document this somewhat. TIA.

File: configure.ac [hunk 3]
```diff
@@ -68,6 +68,23 @@ esac
...
+# _XOPEN_SOURCE_EXTENDED is defined or _XOPEN_SOURCE defined to be
+# >= 500. In DragonFly BSD, FreeBSD and OpenBSD, defining
+# _XOPEN_SOURCE to any value *hides* interfaces that would be useful
+# for htop.
+dnl
+dnl Note: The "#undef" in AH_VERBATIM will be replaced with "#define"
+dnl when config.h is generated.
+AH_VERBATIM([_XOPEN_SOURCE_EXTENDED],
+[/* Enables XPG4v2 (SUSv1) interfaces if they are not enabled already
+   with _XOPEN_SOURCE */
```
Comment (BenBE) [inline configure.ac:81]: Can be one line IMO.

```suggestion
[/* Enables XPG4v2 (SUSv1) interfaces if they are not enabled already with _XOPEN_SOURCE */
```

**Comment by Explorer09** [comment]
@BenBE

I have a idea when trying to improve configure in detecting ncurses header, but am not sure whether such a change is acceptable:

```diff
--- a/ProvideCurses.h
+++ b/ProvideCurses.h
@@ -16,16 +16,13 @@ in the source distribution for its full text.
 
 // IWYU pragma: begin_exports
 
-#if defined(HAVE_NCURSESW_CURSES_H)
-#include <ncursesw/curses.h>
-#elif defined(HAVE_NCURSES_NCURSES_H)
-#include <ncurses/ncurses.h>
-#elif defined(HAVE_NCURSES_CURSES_H)
-#include <ncurses/curses.h>
-#elif defined(HAVE_NCURSES_H)
-#include <ncurses.h>
-#elif defined(HAVE_CURSES_H)
-#include <curses.h>
+// The CURSES_HEADER_NAME macro may be defined by configure script to any of these:
+// <ncursesw/curses.h>
+// <ncurses/curses.h>
+// <ncurses.h>
+// <curses.h>
+#if defined(CURSES_HEADER_NAME)
+#include CURSES_HEADER_NAME
 #endif
 
 #ifdef HAVE_LIBNCURSESW
```

Instead of having a lot of boolean `HAVE_*CURSES_H` macro tokens, I think it would be better to have one token for curses/ncurses header name. This form of include directive is supposed to work with any C89-compliant compiler, but programmers might be unfamiliar with this form of include.

**Comment by BenBE** [comment]
I considered it briefly when that header was created, but decided against it, as we hate the ProvideCurses.h anyway to deal with all the other aspects of curses that come bundled with terminal control stuff. And since you get those booleans for free™ from autoconf there's no real reason not using them in favour of some little-used feature of the pre-processor. So keep it as-is.

**Comment by Explorer09** [comment]
> I considered it briefly when that header was created, but decided against it, as we hate the ProvideCurses.h anyway to deal with all the other aspects of curses that come bundled with terminal control stuff. And since you get those booleans for free™ from autoconf there's no real reason not using them in favour of some little-used feature of the pre-processor. So keep it as-is.

The problem is, the list of possible header names for ncurses is non-exhaustive. We  only check header names that are common in distributions (`<ncursesw/curses.h>`, `<ncurses.h>`, etc.), but I've experimented with ncurses' configure (upstream), and it technically allows builders to specify any suffix for the library and header directory. So the header could be `ncurseswfoo/curses.h` (if ncurses is configured with `--with-extra-suffix=foo`), `ncurseswbar/curses.h` (if `--with-extra-suffix=bar`) and so on.

I know these are custom builds and niche use cases, but as I introduced `--with-curses=XXX` in htop's configure, I think I should allow builders to specify custom header for ncurses (to free them from needing to hack the ProvideCurses.h). It won't be a configure option, but a `CPPFLAGS='-DCURSES_HEADER_NAME=<ncurseswfoo/curses.h>'` override.

**Comment by BenBE** [comment]
People who mess around with the header names may also just mess around with the code at that point. I doubt most programs using (n)curses at that point would even compile properly with these custom header names. That's some additional complexity added that doesn't provide reasonable benefit for the average case or is common enough to warrant going that extra kilometer.

People messing with their header names can feel lucky, that it's basically just one place they have to mess with in our code …

**Comment by Explorer09** [comment]
I think I would defer the header detection changes to another PR. The rest of the changes are ready.

**Pull request merged**